aiohttp/client_middleware_digest_auth.py

```diff
@@ -0,0 +1,415 @@
+
+        # QoP-dependent fields
+        if qop:
+            header_fields["qop"] = qop
+            header_fields["nc"] = ncvalue
+            header_fields["cnonce"] = cnonce
+
+        # Build header using templates for each field type
+        pairs: List[str] = []
+        for field, value in header_fields.items():
+            if field in QUOTED_AUTH_FIELDS:
+                pairs.append(f'{field}="{value}"')
+            else:
+                pairs.append(f"{field}={value}")
+
+        return f"Digest {', '.join(pairs)}"
+
+    def _authenticate(self, response: ClientResponse) -> bool:
+        """
+        Takes the given response and tries digest-auth, if needed.
+
+        Returns true if the original request must be resent.
+        """
+        if response.status != 401:
+            return False
+
+        auth_header = response.headers.get("www-authenticate", "")
+        if not auth_header:
+            return False  # No authentication header present
+
+        method, sep, headers = auth_header.partition(" ")
+        if not sep:
+            # No space found in www-authenticate header
+            return False  # Malformed auth header, missing scheme separator
+
+        if method.lower() != "digest":
+            # Not a digest auth challenge (could be Basic, Bearer, etc.)
+            return False
+
+        if not headers:
+            # We have a digest scheme but no parameters
+            return False  # Malformed digest header, missing parameters
+
+        # We have a digest auth header with content
+        if not (header_pairs := parse_header_pairs(headers)):
+            # Failed to parse any key-value pairs
+            return False  # Malformed digest header, no valid parameters
+
+        # Extract challenge parameters
+        self._challenge = {}
+        for field in CHALLENGE_FIELDS:
+            if value := header_pairs.get(field):
+                self._challenge[field] = value  # type: ignore[literal-required]
+
+        # Return True only if we found at least one challenge parameter
+        return bool(self._challenge)
+
+    async def __call__(
+        self, request: ClientRequest, handler: ClientHandlerType
+    ) -> ClientResponse:
+        """Run the digest auth middleware."""
+        retry_count = 0
+
+        while True:
+            # Apply authorization header if we have a challenge (on second attempt)
+            if retry_count > 0:
+                auth_header = self._encode(request.method, request.url, request.body)
+                if auth_header:
+                    request.headers[hdrs.AUTHORIZATION] = auth_header
+
+            # Send the request
+            response = await handler(request)
+
+            # Check if we need to authenticate
+            if self._authenticate(response) and retry_count < 1:
+                retry_count += 1
+                response.release()  # Release the response to enable connection reuse
```

Comment:
I'd also like to investigate this a bit more. It's a bit awkward if developers are expected to know this.

My understanding is that the connection should already be released if the full response has been read (if under a couple of KB it should have). And if it hasn't been read, then .release() should close the connection anyway...

Reply:
Without the `release()`, it seems to work most of the time. I suspect that's due to a race condition, likely related to [aiohttp issue #10325](https://github.com/aio-libs/aiohttp/issues/10325).

My hunch is that if we address the race properly, `release()` might not be necessary at all.

Reply:
captured in https://github.com/aio-libs/aiohttp/issues/10901